include/exec/__detail/__system_context_default_impl_entry.hpp

```diff
@@ -23,23 +23,111 @@
 
 #include "__system_context_default_impl.hpp"
 
-STDEXEC_PRAGMA_PUSH()
-STDEXEC_PRAGMA_IGNORE_GNU("-Wattributes") // warning: inline function '[...]' declared weak
-
-/// Gets the default system context implementation.
-extern "C" STDEXEC_SYSTEM_CONTEXT_INLINE STDEXEC_ATTRIBUTE((weak))
-__exec_system_context_interface*
-  __get_exec_system_context_impl() {
-  return exec::__system_context_default_impl::__instance_holder::__singleton()
-    .__get_current_instance();
-}
-
-/// Sets the default system context implementation.
-extern "C" STDEXEC_SYSTEM_CONTEXT_INLINE STDEXEC_ATTRIBUTE((weak))
-  void
-  __set_exec_system_context_impl(__exec_system_context_interface* __instance) {
-  return exec::__system_context_default_impl::__instance_holder::__singleton()
-    .__set_current_instance(__instance);
-}
-
-STDEXEC_PRAGMA_POP()
+#include <atomic>
+#include <thread>
+
+namespace exec {
+  namespace __detail {
+    STDEXEC_SYSTEM_CONTEXT_INLINE constinit __spin_lock __system_context_mutex{};
+
+    STDEXEC_SYSTEM_CONTEXT_INLINE constinit std::atomic<system_context_interface*>
+      __system_context_instance{nullptr};
+
+    STDEXEC_SYSTEM_CONTEXT_INLINE constinit std::atomic<new_system_context_handler>
+      __new_system_context_handler{[]() -> system_context_interface* {
+        return new __detail::__system_context_impl{};
+      }};
+
+    // The job of this object is to hold a reference to whatever object
+    // is the current system context from the time the first system
+    // context object is created until after main exits.
+    struct __system_context_keep_alive {
+      __system_context_keep_alive() {
+        (void) get_system_context_instance();
+      }
+
+      ~__system_context_keep_alive() {
+        release_system_context_instance(__system_context_instance.load());
+      }
```

Comment:
If this class were to actually be used, I would have concerns about it potentially incrementing the ref-count on one system-context object and then decrementing the ref-count on another system-context object if anyone ever called `set_new_system_context_handler` during the lifetime of this object.

Reply:
Also, if the intention is that the `__system_context_base` is intended to be a base-class of the system-context object then having the `__system_context_keep_alive` constructor call `get_system_context_instance()` is going to end up always calling the system-context construction function twice.

The first cal to `get_system_context_instance()` by the user is going to construct the default system-context, then the constructor of that system-context object is going to call `get_system_context_instance()` again, which is then going to try to construct a second system-context object as we're not yet finished constructing the first one, which is going to then call into the constructor of the `__system_context_base` again.

Best case, this will redundantly construct two system-context objects, install one and then have the second one replace the first, destroying the first. Worst-case, you'll hit a terminate() because you are reentrantly trying to initialize a function-local static object.

Reply:
> I would have concerns about it potentially incrementing the ref-count on one system-context object and then decrementing the ref-count on another system-context object

that is precisely the design. `set_new_system_context_handler` is responsible for fixing up the ref-counts when the system context object changes. i've said this several times already.